builtin/providers/azure/config.go

```diff
@@ -0,0 +1,63 @@
+package azure
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/management"
+)
+
+// Config is a struct which holds all the required information to access
+// Azure services.
+type Config struct {
+	PublishSettingsFilePath string
+	SubscriptionID          string
+	ManagementCert          []byte
+	ManagementUrl           string
+}
+
+// AzureClient contains all the handles required for managing Azure services.
+type AzureClient struct {
+	// the client which holds the authentification information
+	managementClient management.Client
+
+	// unfortunately; because of how Azure's network API works; doing networking operations
+	// concurrently is very hazardous, and we need a mutex.
+	mutex *sync.Mutex
```

Comment:
I actually don't think this gives you the behaviour you would expect. Have a look at the Terraform internals around plugin communication to see for yourself, but I'll think you'll agree with me...

Once we loose the mutex here, having a custom `AzureClient` doesn't make sense anymore, so we could stick to just `management.Client` instead.